beacon_node/eth1/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+    /// From https://github.com/ethereum/eth2.0-specs/blob/v0.8.1/specs/validator/0_beacon-chain-validator.md#eth1-data
+    pub fn get_eth1_votes<T: EthSpec>(
+        &self,
+        state: &BeaconState<T>,
+        previous_eth1_distance_hash: H256,
+    ) -> Result<Eth1Data> {
+        // TODO: Need a better way to get `previous_eth1_distance`.
+        let previous_eth1_distance = tokio::runtime::current_thread::block_on_all(
+            self.fetcher
+                .get_block_height_by_hash(previous_eth1_distance_hash),
+        )?;
+        let previous_eth1_distance =
+            U128::as_u64(&previous_eth1_distance.ok_or(Error::InvalidParam)?);
+        let new_eth1_data = self
+            .eth1_data_cache
+            .get_eth1_data_in_range(ETH1_FOLLOW_DISTANCE, 2 * ETH1_FOLLOW_DISTANCE);
+        let all_eth1_data = self
+            .eth1_data_cache
+            .get_eth1_data_in_range(ETH1_FOLLOW_DISTANCE, previous_eth1_distance);
+        let mut valid_votes: Vec<Eth1Data> = vec![];
+        for (slot, vote) in state.eth1_data_votes.iter().enumerate() {
+            let period_tail: bool = (slot as u64 % T::SlotsPerEth1VotingPeriod::to_u64())
+                >= ((T::SlotsPerEth1VotingPeriod::to_u64() as f64).sqrt() as u64 + 1);
+            if new_eth1_data.contains(vote) || (period_tail && all_eth1_data.contains(vote)) {
+                valid_votes.push(vote.clone());
+            }
+        }
+        Ok(valid_votes
+            .iter()
+            .cloned()
+            .max_by(|x, y| {
+                let mut result = valid_votes
+                    .iter()
+                    .filter(|n| *n == x)
+                    .count()
+                    .cmp(&valid_votes.iter().filter(|n| *n == y).count());
+                if result == Ordering::Equal {
+                    result = all_eth1_data
+                        .iter()
+                        .position(|s| s == x)
+                        .cmp(&all_eth1_data.iter().position(|s| s == y));
+                }
+                result
+            })
+            .unwrap_or(self.eth1_data_cache.get_eth1_data(ETH1_FOLLOW_DISTANCE)?))
+    }
+}
+
+pub fn run<F: Eth1DataFetcher + 'static>(
+    eth1: Eth1<F>,
+    executor: &TaskExecutor,
+    log: &slog::Logger,
+) {
+    let log = log.new(o!("service" => "eth1_chain"));
+
+    // Run a task for calling `update_cache` periodically.
+    let eth1_block_time_seconds: u64 = 5; // Approximate block time for eth1 chain.
+    let eth1_block_interval: u64 = 3; // Interval of eth1 blocks to update eth1_data_cache.
+    let interval_log = log.clone();
+    let eth1_data_interval = {
+        // Set the interval to fire every `eth1_block_interval` blocks.
+        let update_duration = Duration::from_secs(eth1_block_interval * eth1_block_time_seconds);
+        Interval::new(Instant::now(), update_duration)
+            .map_err(move |_| warn!(interval_log, "Interval timer failing"))
```

Comment:
It would be nice to include the error with this log. E.g.,

```rust
.map_err(move |e| {
  warn!(
    interval_log,
    "Interval timer failing";
    "error" => format!("{:?}", e)
  )
)
```

Reply:
Thanks. This is much better.